src/purview.ts

```diff
@@ -326,11 +324,19 @@ export function handleWebSocket(
 
     ws.on("close", async () => {
       wsState.closing = true
-      if (!wsState.connected) {
+      // We check if `connectionState` is set to "connected" because it could be
+      // the case that the "close" event fires just after the "connect" event
+      // (e.g., on page refresh), and the "close" event will see that the
+      // `wsState.roots` is an empty array due to the "connect" still being in
+      // progress. This would result in an incomplete cleanup of the previous
+      // connection's state. Hence, we return early, set the `closing` flag, and
+      // let the "connect" event clean up the existing state by signaling with
+      // `closing`.
+      if (wsState.connectionState !== "connected") {
         return
       }
 
-      await saveRootsState(wsState)
+      await cleanupExistingState(wsState)
```

Comment:
as a verb, clean up is two words. I think it's implied the state exists if you're cleaning it up, so I'd suggest e.g. `cleanUpWebSocketState`